### PR TITLE
Switch to rhel-coreos-9

### PR DESCRIPTION
**Pull request by cgwalters**
daemon: Clean up `switchKernel` a bit

De-duplicate calls to `canonicalizeKernelType` to make the
logic easier to read.  Also add a few comments.

---

vendor: Bump coreos/rpm-ostree-client-go

In prep for usage in MCD.

---

daemon: Make switchKernel less stateful

This is prep for fixing RHEL9 upgrades while maintaining `kernel-rt`.

Previously the `switchKernel` logic tried to carefully handle
all 4 cases (default -> default, default -> rt, rt -> default, rt -> rt).

But, the last one (rt -> rt) was not quite right because
the previous `rpm-ostree rebase` command already preserved the previous
kernel.  In fact it was pretty expensive to do things this way
because we'd e.g. regenerate the initramfs *twice*.

To say this another way: when doing a RHEL9 update, it's actually
the first `rpm-ostree rebase` command which fails before we
even get to `switchKernel`.

And the reason is due to the introduction of a new `-core` subpackage;
xref https://issues.redhat.com/browse/OCPBUGS-8113

So here's the new logic to handle this:

- Before we do the `rebase` operation to the new OS, we detect
  any previous overrides of any packages starting with `kernel-rt`
  and we remove them.  Notably this avoids hardcoding any specific
  kernel subpackages; we just remove *everything* starting with
  `kernel-rt` which should be more robust to subpackage changes
  in the future.
- Consequently the `rebase` operation will hence start out by deploying the
  stock image i.e. with throughput kernel (though note we *are*
  carefully preserving other local overrides)
- The `switchKernel` function now longer needs to take the *previous*
  machineconfig state into account (except for logging).
  Instead, we just detect if the target is RT, and if so we then we
  apply the latest packages.

This significantly simplifies the logic in `switchKernel`, and will
help fix RHEL9 upgrades.

---

Switch to rhel-coreos-9

ref: https://issues.redhat.com/browse/COS-1983

We introduced a new `rhel-coreos-9` to aid having a switch be
an atomic operation.

---

daemon: Also override `kernel-modules-core`

Unfortunately rpm-ostree requires this right now; we have an issue
and code to provide a better API in https://github.com/coreos/rpm-ostree/issues/2542
But using that will require shipping the updated rpm-ostree in RHEL 8.6.z
or at least OCP 4.12.z, which is problematic.

Because we know the new MCD will always be upgrading to RHEL9,
for now let's update this hardcoded list.  In the future we can
detect when the running host has `--remove-installed-kernel` and
use it instead.

---

openshift-azure-routes: Avoid synchronizing too quickly

Rapid file changes triggering the path unit can start the
service here frequently, and then this can cause the start
limit to be hit, and then systemd will refuse further
activations (unless we bumped the limit).

I don't think we need to synchronize the iptables
rules more than once every 3 seconds.

---

ensures that RHCOS 9 SSH keys are in the right place

---

OKD release controller is out-of-date

---

ensures SSH keys get moved to the correct location

When we move from RHCOS 8 -> RHCOS 9, the SSH keys are not being written
to the new location because:

1. When the upgrade configs are written to the node, it is still running RHCOS 8, so the keys are not being written to the new location.
2. The node reboots into RHCOS 9 to complete the upgrade.
3. The "are we on the latest config" functions detect that we are indeed on the latest config and so it does not attempt to perform an update.

---

teaches TestIgn3Cfg about the new RHCOS 9 key path

---

### CODE REVIEW

**Comment by cgwalters** [comment]
Interesting...in the GCP run, the workers came up fine as rhel9, but the control plane stayed rhel8?  https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3485/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1613636820694732800/artifacts/e2e-gcp-op/gather-extra/artifacts/oc_cmds/nodes

**Comment by cgwalters** [comment]
Oh duh I see, need to change the bootstrap path too.

/test all

**Comment by cgwalters** [comment]
/test e2e-gcp-op
:arrow_up: this one only failed in deprovisioning AFAICS!

```
NAME                                        STATUS   ROLES                  AGE    VERSION           INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                                       KERNEL-VERSION                CONTAINER-RUNTIME
ci-op-qkwynd52-1354f-df8wh-master-0         Ready    control-plane,master   140m   v1.25.2+0003605   10.0.0.4      <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
ci-op-qkwynd52-1354f-df8wh-master-1         Ready    control-plane,master   140m   v1.25.2+0003605   10.0.0.5      <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
ci-op-qkwynd52-1354f-df8wh-master-2         Ready    control-plane,master   138m   v1.25.2+0003605   10.0.0.3      <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
ci-op-qkwynd52-1354f-df8wh-worker-a-w7vxv   Ready    worker                 123m   v1.25.2+0003605   10.0.128.3    <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
ci-op-qkwynd52-1354f-df8wh-worker-b-ps97v   Ready    worker                 123m   v1.25.2+0003605   10.0.128.4    <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
ci-op-qkwynd52-1354f-df8wh-worker-c-62c72   Ready    worker                 123m   v1.25.2+0003605   10.0.128.2    <none>        Red Hat Enterprise Linux CoreOS 413.90.202212151724-0 (Plow)   5.14.0-70.36.1.el9_0.x86_64   cri-o://1.25.0-53.rhaos4.12.git2002c49.el9
```


The e2e-aws-ovn test...
"[sig-node] Ephemeral Containers [NodeConformance] will start an ephemeral container in an existing pod"
sure looks like an existing flake.
/test e2e-aws-ovn

Hypershift...dunno?

But let's do this a bit more seriously:
/payload 4.13 ci blocking

**Comment by cgwalters** [comment]
Looks like at least one SDN pod is trying to execute the host's copy of `oc` in the container userspace, which won't work since it now needs a newer glibc:

>  /host/usr/bin/oc: /lib64/libc.so.6: version `GLIBC_2.33' not found (required by /host/usr/bin/oc)

**Comment by dgoodwin** [comment]
> @cgwalters: trigger 4 job(s) of type blocking for the ci release of OCP 4.13
> 
>     * periodic-ci-openshift-release-master-ci-4.13-e2e-aws-ovn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-aws-ovn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-azure-sdn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-e2e-aws-sdn-serial
> 
> 
> See details on https://pr-payload-tests.ci.openshift.org/runs/ci/9a3c9740-92ed-11ed-97df-8ca781b2fb9f-0

Just wanted to help with the results here, it's awesome that we can run this now on a new OS image pre-merge. The payload is showing some clear regressions specifically with disruption related to ingress services, oauth and console.

ingress-to-oauth-server-new-connections was unreachable during disruption testing for at least 8m36s of 1h28m43s (maxAllowed=10s)

This is showing up in all 10 runs on 
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-e2e-aws-ovn-upgrade/1613731214009569280 (aws ovn micro upgrade)

For those unaware, for these disruptions a good starting point is to expand the intervals chart on the prow page (usually the first one), you will see when the disruption happened in time and be able to correlate that with whatever else was going on in the cluster.

[sig-node] Ephemeral Containers [NodeConformance] will start an ephemeral container in an existing pod [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
- this failed on all ten runs for:
  - https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-e2e-aws-ovn-upgrade/1613731214009569280
  - https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-aws-ovn-upgrade/1613731219789320192
  
Upgrade is not succeeding at all on Azure but I heard that may already be known: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-azure-sdn-upgrade/1613731229264252928

And "payload 4.13 nightly blocking" will also give some good coverage, different combos than CI, and a little more of them.

**Comment by cgwalters** [comment]
Just to get some additional coverage
/test e2e-aws-single-node
/test e2e-gcp-op-single-node
/test e2e-metal-assisted

**Comment by cgwalters** [comment]
> ingress-to-oauth-server-new-connections was unreachable during disruption testing for at least 8m36s of 1h28m43s (maxAllowed=10s)

I don't personally have a lot of expertise on this.

**Comment by cheesesashimi** [comment]
Running the FIPS and RT MCO jobs:

/test e2e-aws-ovn-fips
/test e2e-aws-ovn-fips-op
/test e2e-gcp-rt
/test e2e-gcp-rt-op

Note: The `e2e-gcp-rt-op` job is expected to fail with either (or both) `TestKernelArguments` and `TestKernelType` tests failing. I've opened https://github.com/openshift/machine-config-operator/pull/3504 to green those up.

**Comment by cheesesashimi** [comment]
Interesting. The new jobs aren't showing up here yet.

**Comment by dgoodwin** [comment]
> > ingress-to-oauth-server-new-connections was unreachable during disruption testing for at least 8m36s of 1h28m43s (maxAllowed=10s)
> 
> I don't personally have a lot of expertise on this.

Few do, we may need the Ingress or SDN teams assistance to debug. But following the instructions I gave above, the outage appears to coincide with OVN alerts like "NoOvnRunningMaster", RouteHealth alerts for console and oauth, ovn-raft-quorum-guard PodDisruptionBudget, ovnkube-node target down, cloud network controller not scheduled on any nodes. It looks more like network than an ingress problem for sure.

Additionally and separately, this test seems to be failing broadly: [sig-node] Ephemeral Containers [NodeConformance] will start an ephemeral container in an existing pod [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]

**Comment by cheesesashimi** [comment]
My findings and analysis for the additional CI jobs I've added can be found here: https://issues.redhat.com/browse/COS-1990

**Comment by sdodson** [comment]
/payload 4.13 nightly blocking
/payload 4.13 ci blocking

Ok, got fresh RHCOS 413.92.202302081904-0 tagged into the image streams referenced by this PR. This should include Kube 1.26 rebase kubelet and everything.

**Comment by dgoodwin** [comment]
> @sdodson: trigger 6 job(s) of type blocking for the nightly release of OCP 4.13
> 
>     * periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-sdn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-e2e-azure-ovn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade
> 
>     * periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-sdn-serial
> 
>     * periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-ovn-ipv6
> 
>     * periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-sdn
> 
> 
> See details on https://pr-payload-tests.ci.openshift.org/runs/ci/66c80230-a82a-11ed-942b-09a9c24932a7-0
> 
> trigger 4 job(s) of type blocking for the ci release of OCP 4.13
> 
>     * periodic-ci-openshift-release-master-ci-4.13-e2e-aws-ovn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-aws-ovn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-azure-sdn-upgrade
> 
>     * periodic-ci-openshift-release-master-ci-4.13-e2e-aws-sdn-serial
> 
> 
> See details on https://pr-payload-tests.ci.openshift.org/runs/ci/66c80230-a82a-11ed-942b-09a9c24932a7-1

First round on analysis on this: https://docs.google.com/document/d/1ObAO0jaRNBeTbHTl0_rNawe7E36P99ifJH0rCOlGpt4/edit?usp=sharing

TL;DR: machine-config operator can't upgrade

**Comment by cgwalters** [comment]
I have no idea what's going wrong in the metal jobs.  I think the failure scenario here is that the control plane nodes are failing to come up - but here on packet.net we aren't gathering console logs.

**Comment by dgoodwin** [comment]
> I have no idea what's going wrong in the metal jobs. I think the failure scenario here is that the control plane nodes are failing to come up - but here on packet.net we aren't gathering console logs.

I'll get them a bug and find someone who can look.

**Comment by cheesesashimi** [comment]
Let's run some of the more uncommon test jobs:

**MCO-specific e2e tests:**
/test e2e-gcp-op
/test e2e-gcp-op-single-node

**OpenShift e2e tests on FIPS:**
/test e2e-aws-ovn-fips

**OpenShift e2e tests on realtime kernel:**
/test e2e-gcp-rt

**Comment by cgwalters** [comment]
https://issues.redhat.com/browse/OCPBUGS-7275 should be fixed :pray: 
/payload 4.13 ci blocking

**Comment by cgwalters** [comment]
OK I think to get the metal coverage we also need
/payload 4.13 nightly blocking

**Comment by cgwalters** [comment]
(looks like what's in 4.13-9.2 is 6 days old; not too bad but may be worth doing a retag if we see any failures that we think were fixed since then)

**Comment by sdodson** [comment]
/payload 4.13 nightly blocking
The 4.12 build with the rpm-ostree fixes was accepted just after midnight last night. This may require manual testing via cluster bot to verify the upgrade is successful as I think the "stable" in the job name means it starts with the latest z-stream.

**Comment by sdodson** [comment]
> (looks like what's in 4.13-9.2 is 6 days old; not too bad but may be worth doing a retag if we see any failures that we think were fixed since then)

We got a new centos stream compose, it'd be good to get a new rhcos build today with that new content and see where we get to.

**Comment by cgwalters** [comment]
In the GCP failures there it looks like it didn't pull in the latest CI 4.12 build; still had the old rpm-ostree.

**Comment by sdodson** [comment]
> In the GCP failures there it looks like it didn't pull in the latest CI 4.12 build; still had the old rpm-ostree.

Yeah, I believe unfortunately this test relies on the latest GA z-stream. I'll work through a cluster-bot upgrade from the latest ci payload and we'll call this one verified once I can confirm that. Unfortunately that won't land in the payload jobs until Jan 27th unless we go and change those CI jobs to start from the latest CI payload.

**Comment by cgwalters** [comment]
> I'll work through a cluster-bot upgrade from the latest ci payload and we'll call this one verified once I can confirm that.

Me too, I did a `launch registry.ci.openshift.org/ocp/release:4.13.0-0.ci-2023-02-16-135040 gcp,rt` and will upgrade to this image.

**Comment by cgwalters** [comment]
Something still a little ill-defined here is mechanically: Do we move forward with effectively merging this PR as is?  That would mean we continue with the new `rhel-coreos-9` image, and drop `rhel-coreos-8`.

**Comment by dgoodwin** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-upgrade

Just looking to get some signal on non-realtime gcp, we're in the dark somewhat so far.

**Comment by sdodson** [comment]
/payload 4.13 nightly blocking
413.92.202302161334-0 has been tagged in

**Comment by sdodson** [comment]
@cgwalters Actually, that may be what we get from periodic-ci-openshift-release-master-ci-4.13-e2e-azure-ovn-upgrade
 but won't that have rpmdb already in the desired format? I don't fully understand the bug being fixed.

**Comment by cgwalters** [comment]
Hmm, I'm not sure why e2e-gcp-op is failing in `testKernelType`.  Not seeing anything obvious in the MCD logs.

**Comment by sdodson** [comment]
/payload 4.13 nightly blocking
/payload 4.13 ci blocking
413.92.202302171914-0 tagged in

**Comment by jcpowermac** [comment]
e2e-vsphere job failure caused by:
https://issues.redhat.com/browse/OCPBUGS-7616
otherwise looks ok

**Comment by dgoodwin** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-ovn-ipv6
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-sdn

Funky results on these last run, we'd like to see what happens next time.

**Comment by dgoodwin** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-ovn-ipv6
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-metal-ipi-sdn

**Comment by sdodson** [comment]
/payload 4.13 nightly blocking
/payload 4.13 nightly informing
NFS tests should be skipped now.

**Comment by sdodson** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-gcp-ovn-rt

**Comment by dgoodwin** [comment]
@sdodson we're in the middle of payloads but I'll post a mix of what's completed in the recent set so far, vs the previous batch of blocking. metal-ipi-sdn just cleared up so I think the previous result was a fluke, bunch of storage tests failed. 

The last nightly azure aggregated jobs showed a trend of disruption, unless we see it again I think this is a fluke, which is not uncommon.

The last aws ovn aggregated job fluked out on ci problems, I think this has passed recently, https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-aws-ovn-upgrade/1629682395848380416

The last azure ci aggregated job looks like at least some NFS test ran and nodes died again, https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/aggregator-periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-azure-sdn-upgrade/1629682405088432128. Maybe we didn't get all tests disabled?

And of course gcp-rt still failing.

Otherwise as far as blockers go, we're looking good I think.

**Comment by sdodson** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-gcp-ovn-rt

**Comment by dgoodwin** [comment]
/payload 4.13 ci blocking

Apologies I set the precedent with a typo that ran 4.12, we want 4.13.

**Comment by dgoodwin** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.13-e2e-azure-ovn-upgrade 10

**Comment by sdodson** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

**Comment by dgoodwin** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.13-e2e-azure-ovn-upgrade 10

**Comment by dgoodwin** [comment]
/payload 4.13 nightly blocking

CI broke last attempt.

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -1163,7 +1163,8 @@ func (dn *CoreOSDaemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig)
 		return nil
 	}
 
-	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-extra"}
+	// TODO: Drop this code and use https://github.com/coreos/rpm-ostree/issues/2542 instead
+	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-core", "kernel-modules-extra"}
```
Comment (cheesesashimi) [inline pkg/daemon/update.go:1167]: question: Will this work for upgrade cases too?

I think this might only work for new-install cases. For upgrade cases (e.g., someone upgrading from 4.12 to 4.13, which will imply RHCOS 8.6 -> 9.0+), we might need to do something like:

```golang
if dn.os.IsEL() && !dn.os.IsEL9() && dn.incomingOS.IsEL9() {
  // ...
}
```

Comment (cgwalters) [reply pkg/daemon/update.go:1167]: I'm aiming to fix the upgrade case i.e. 4.12 to 4.13.  Is the other case you're thinking of 4.13.x to 4.13.y?  That should also work because the relevant kernel package is present there too.

Another way to say this is: Here we're relying on the fact that this change is *only* in the 4.13 MCD, and when we're running the 4.13 MCO we know (currently) that we're going to be upgrading to a RHEL9 era OS.

Comment (cgwalters) [reply pkg/daemon/update.go:1167]: The overall code here is also currently gated on

```
	// We support Kernel update only on RHCOS and SCOS nodes
	if !dn.os.IsEL() {
		glog.Info("updating kernel on non-RHCOS nodes is not supported")
		return nil
	}
```

(There's no kernel variants in Fedora today, but if the rt kernel ever fully merges, there probably will be)

Comment (cheesesashimi) [reply pkg/daemon/update.go:1167]: I was thinking of the 4.13.x -> 4.13.y case because of the OS upgrade that occurs during installation. In other words: When this is installed on a new cluster, the cluster is initially brought up using the RHCOS 8.6 boot images, then eventually layers the RHCOS 9 container image onto the node. That means the MCD will run in both RHCOS 8.6 as well as RHCOS 9.

It sounds like the 4.12 -> 4.13 case is pretty clear. Although I wonder how that would work in the RT kernel case since the package names have changed.

Comment (cgwalters) [reply pkg/daemon/update.go:1167]: :tada: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3485/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1631289321837301760/artifacts/e2e-gcp-op/test/build-log.txt has `--- PASS: TestKernelType (625.74s)`

**Comment by cgwalters** [comment]
But also that job has
```
    mcd_test.go:431: updated ssh keys not found in authorized_keys, got XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
--- FAIL: TestNoReboot (48.32s)
```

cc @cheesesashimi

**Comment by cheesesashimi** [comment]
The `TestNoReboot` failure is most likely because https://github.com/openshift/machine-config-operator/pull/3534 hasn't landed yet.

File: pkg/daemon/update.go (hunk 1 above)
Comment (cgwalters) [reply pkg/daemon/update.go:1167]: > I was thinking of the 4.13.x -> 4.13.y case because of the OS upgrade that occurs during installation. In other words: When this is installed on a new cluster, the cluster is initially brought up using the RHCOS 8.6 boot images, then eventually layers the RHCOS 9 container image onto the node. That means the MCD will run in both RHCOS 8.6 as well as RHCOS 9. Although this might only be an issue if someone configured their cluster on Day 1 with an RT kernel instead of doing it as a Day 2 operation.

Hmm...but that's only a transient situation for now, we're going to land the PR https://github.com/openshift/installer/pull/6886 to bump it to 9.2 as well for 4.13.

IOW, clusters born-in 4.13 will start from RHEL9 too.

But also even in the situation today, I believe what happens if you provide a "day 0" machineconfig with `kerneltype: realtime`, we'll still do the OS update and then kernel switch as a *transactional* operation (because of how rpm-ostree works - changes always operate on the pending deployment).  

IOW the MCO is doing
```
rpm-ostree rebase <rhel9>
rpm-ostree switch-to-kernel-rt
```

And even if the running host is RHEL8 (as it is today) that operation will always run targeting the RHEL9 OS.

**Comment by cgwalters** [comment]
> The TestNoReboot failure is most likely because https://github.com/openshift/machine-config-operator/pull/3534 hasn't landed yet.

OK, I've merged that PR into this PR, and I also pushed a hack-around for having the `azure-vip-route.service` hit its start limit, which I still think is involved in https://issues.redhat.com/browse/TRT-876 even if evidence is leading us elsewhere.

**Comment by cheesesashimi** [comment]
> OK, I've merged that [[#3534](https://github.com/openshift/machine-config-operator/pull/3534)] PR into this PR

question: Do you want to try to land this with that PR or individually? I'm fine with either.

File: pkg/daemon/update.go (hunk 1 above)
Comment (cheesesashimi) [reply pkg/daemon/update.go:1166]: Got it. That makes sense!

**Comment by cgwalters** [comment]
> question: Do you want to try to land this with that PR or individually? I'm fine with either.

I don't have a strong opinion, it's more up to QE right that we're blocking on for merging that now AIUI?  Which is easier to test?  Since we only need it with rhel9, perhaps easier to verify all at once in this PR/

```
=== RUN   TestIgn3Cfg
    mcd_test.go:786: Created 99-ign3cfg-infra-d73fe413-ac41-4ba3-ad5f-f62686a0d655
    utils.go:173: Pool infra has rendered configs [99-ign3cfg-infra-d73fe413-ac41-4ba3-ad5f-f62686a0d655] with rendered-infra-4aa90237a7aca7b03956e2b685ef93b6 (waited 6.085694697s)
    utils.go:205: Pool infra has completed rendered-infra-4aa90237a7aca7b03956e2b685ef93b6 (waited 2m50.040689939s)
grep: /rootfs/home/core/.ssh/authorized_keys: No such file or directory
command terminated with exit code 2
    mcd_test.go:799: 
        	Error Trace:	/go/src/github.com/openshift/machine-config-operator/test/e2e/utils.go:462
        	            				/go/src/github.com/openshift/machine-config-operator/test/e2e/mcd_test.go:799
        	Error:      	Expected nil, but got: &exec.ExitError{ProcessState:(*os.ProcessState)(0xc000013bf0), Stderr:[]uint8(nil)}
        	Test:       	TestIgn3Cfg
        	Messages:   	failed to exec cmd [grep 1234_test_ign3 /rootfs/home/core/.ssh/authorized_keys] on node ci-op-h5wxn42l-1354f-rmq56-worker-c-xztnc: 
--- FAIL: TestIgn3Cfg (177.12s)
```

**Comment by cgwalters** [comment]
Another run, I am curious if my azure-routes change helped anything or not
/payload 4.13 nightly blocking

**Comment by cheesesashimi** [comment]
Ahh, I know why that test failed. It never learned about the location of the new RHCOS 9 key path. I've pushed a fix to my PR branch that you can pull in. As an aside, this is yet another one of those cases where I find myself longing for a better MCD testing story.

**Comment by sdodson** [comment]
/payload 4.13 ci informing
/payload 4.13 ci blocking
/payload 4.13 nightly informing
rounding out the payload jobs that haven't been run recently

**Comment by cgwalters** [comment]
OK so we fixed day 0 kernel-rt installs, but actually we're still broken when upgrading from 4.12.z RT to 4.13 here ([logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/openshift-machine-config-operator-3485-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade/1631435509865648128/artifacts/e2e-gcp-ovn-rt-upgrade/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-daemon-znkrd_machine-config-daemon.log)):

> E0303 01:31:36.097915   78327 writer.go:200] Marking Degraded due to: failed to update OS to registry.build01.ci.openshift.org/ci-op-kcmkpgq1/stable@sha256:ab2fa6f321f12af1e45f19d928199aa7fc4a6341aa77d52607bdab7d93ba130b : error running rpm-ostree rebase --experimental ostree-unverified-registry:registry.build01.ci.openshift.org/ci-op-kcmkpgq1/stable@sha256:ab2fa6f321f12af1e45f19d928199aa7fc4a6341aa77d52607bdab7d93ba130b: [0m[31merror: [0mCould not depsolve transaction; 1 problem detected:
 Problem: package kernel-modules-core-5.14.0-282.el9.x86_64 requires kernel-uname-r = 5.14.0-282.el9.x86_64, but none of the providers can be installed
  - conflicting requests
: exit status 1

In this scenario...ugh, it's messy.  The kernel type isn't changing, and the current MCD code here is relying on "rpm-ostree rebase" to just DTRT.  Handling this without https://github.com/coreos/rpm-ostree/pull/4322 is in this case really going to force us into OS conditionals unfortunately.

**Comment by cheesesashimi** [comment]
If we need to consider OS conditionals, particularly those of the incoming container, I have https://github.com/openshift/machine-config-operator/pull/3557 currently inflight. This is the "infer OS version from OS container image labels" code.

**Comment by cgwalters** [comment]
> If we need to consider OS conditionals, particularly those of the incoming container, I have https://github.com/openshift/machine-config-operator/pull/3557 currently inflight. This is the "infer OS version from OS container image labels" code.

Hmm...thinking about this...I think I said the problem incorrectly.  We always know that the incoming container today will be RHEL9 (right¹?).  I think this problem isn't so much about detecting, it's that fundamentally today the MCD's logic must change to handle the upgrade case here because we can't just call `rpm-ostree upgrade` and expect the previous overrides to apply.

The hardcoded package lists in `defaultKernel` and `realtimeKernel` are the real problem here - and what's causing this problem is that we're trying to keep the rhel8 era overrides in place when doing a rhel9 update.

I did https://github.com/coreos/rpmostree-client-go/pull/13 which will allow us to query the previous state, although I'm still thinking of potentially easier ways out of this.



¹ Though clearly it'd be useful to still e.g. support someone doing an OS override back to rhel8 in 4.13 at least for testing/comparison purposes for a while, so it'd clearly be *better* if we more gracefully handled all of this consistently and actually didn't always assume the incoming OS was rhel9.

**Comment by cheesesashimi** [comment]
> I think this problem isn't so much about detecting, it's that fundamentally today the MCD's logic must change to handle the upgrade case here because we can't just call rpm-ostree upgrade and expect the previous overrides to apply.

Ahh OK. Now I understand the problem better, thanks for the clarification.

**Comment by cgwalters** [comment]
OK I pushed a split-out PR in https://github.com/openshift/machine-config-operator/pull/3580 which attempts to fix this - and hopefully we'll get signal that that PR doesn't regress things on the current RHEL8 scenarios and can land it in master (today?).

But I've now merged that PR into this one too, and let's see if my compile-tested-only code works with
/test e2e-gcp-ovn-rt-upgrade

**Comment by sdodson** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-alibaba-ovn

**Comment by cgwalters** [comment]
Hmm it seems like the image job is cached or something?  Pushed another commit which I think will fix the previous commit, but also hopefully blow out the cache.

**Comment by cheesesashimi** [comment]
I'm seeing that in a few other places as well.

**Comment by cgwalters** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

**Comment by cgwalters** [comment]
> /payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

BTW let's all try to collectively indelibly etch this one into our brains - this is the upgrade test we want to run when we're making changes to OS upgrades like this and especially RT kernel stuff.  (the kernel test in `e2e-gcp-op` is good, but this is better)

**Comment by cgwalters** [comment]
OK so https://github.com/openshift/machine-config-operator/pull/3580 is good for master (which is still rhel8.6 -> rhel8.6).  Let's test things out on this path and see whether it works here too!
/test e2e-gcp-op
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

**Comment by cgwalters** [comment]
Yep, this works here too :tada: 

e2e-gcp-op is green, and the payload job is good:

```
Deployments:
* ostree-unverified-registry:registry.build01.ci.openshift.org/ci-op-jky2jhft/stable@sha256:ab2fa6f321f12af1e45f19d928199aa7fc4a6341aa77d52607bdab7d93ba130b
                   Digest: sha256:ab2fa6f321f12af1e45f19d928199aa7fc4a6341aa77d52607bdab7d93ba130b
                  Version: 413.92.202303011445-0 (2023-03-05T01:05:59Z)
      RemovedBasePackages: kernel-core kernel-modules kernel kernel-modules-core kernel-modules-extra 5.14.0-282.el9
          LayeredPackages: kernel-rt-core kernel-rt-kvm kernel-rt-modules
                           kernel-rt-modules-extra

  (error fetching image metadata)
                Timestamp: 2023-03-04T23:38:13Z
      RemovedBasePackages: kernel-core kernel-modules kernel kernel-modules-extra 4.18.0-372.46.1.el8_6
          LayeredPackages: kernel-rt-core kernel-rt-kvm kernel-rt-modules
                           kernel-rt-modules-extra
```

**Comment by sdodson** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.13-e2e-alibaba-ovn

**Comment by cgwalters** [comment]
Rebased on the latest https://github.com/openshift/machine-config-operator/pull/3580
Also lifting draft since we have to merge this soon

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
At first-glance, you'll need to re-integrate https://github.com/openshift/machine-config-operator/pull/3534 because QE found some issues with the directory permissions for `/home/core/.ssh/authorized_keys.d`, which I've since fixed.

The short version is that since the MCO shells out to `mkdir -p` instead of calling `os.MkdirAll()`, only the inner-most directory gets created with the desired permissions (`0700` in this case). The rest are created with `0755`.
